java/src/main/native/src/TableJni.cpp

```diff
@@ -1037,9 +1037,9 @@ cudf::io::schema_element read_schema_element(int& index,
     // go to the next entry, so recursion can parse it.
     index++;
     for (int i = 0; i < num_children; i++) {
+      auto const name = std::string{names.get(index).get()};
       child_elems.insert(
-        std::pair{names.get(index).get(),
-                  cudf::jni::read_schema_element(index, children, names, types, scales)});
+        std::pair{name, cudf::jni::read_schema_element(index, children, names, types, scales)});
```

Comment:
Notice that the value `index` is modified inside `read_schema_element` thus `name` must be evaluated before calling to `insert`.

I'm surprised that this was not discovered before. I copied this code into a new place and it immediately triggered out of bound access.